apis/python/src/tiledbsoma/util.py

```diff
@@ -127,3 +129,52 @@ def slice_to_range(ids: slice) -> Optional[Tuple[int, int]]:
     if ids.step is not None and ids.step != 1:
         raise ValueError("slice step must be 1 or None")
     return (ids.start, ids.stop)  # XXX TEMP
+
+
+def dense_indices_to_shape(
+    coords: Tuple[Union[int, slice], ...],
+    array_shape: Tuple[int, ...],
+    result_order: ResultOrder,
+) -> Tuple[int, ...]:
+    """
+    Given a subarray index specified as a tuple of per-dimension slices or scalars
+    (eg, ``([:], 1, [1:2])``), and the shape of the array, return the shape of
+    the subarray.
+
+    See read_tensor for usage.
+    """
+    shape: List[int] = []
+    for n, coord in enumerate(coords):
+        shape.append(dense_index_to_shape(coord, array_shape[n]))
```

Comment:
A comprehension with `zip` might be more natural here:

```suggestion
    shape = [
        dense_index_to_shape(coord, dim_shape)
        for (coord, dim_shape) in zip(coords, array_shape)
    ]
```